tools/validators/instance_validator/validate/entity_instance.py

```diff
@@ -125,21 +125,77 @@ def _ValidateLinks(self, entity_instances):
+      target_entity_type = self.universe.GetEntityType(self.namespace,
+                                                       self.entity_type)
+      all_fields_dict = src_entity_type.GetAllFields().copy()
+      for source_field, target_field  in fields_map.items():
+        opt_wrapper_field = all_fields_dict.pop('/'+source_field.data, None)
+        if opt_wrapper_field is None:  # an extra field that should not be here
+          print('Invalid extra link present:', source_field.data)
           return False
-
+        # check the fields are present
+        # assumes that the namespace is `` for now
+        if not src_entity_type.HasField('/'+source_field.data):
+          print('Invalid links field source: ', source_field)
+          return False
+        if not target_entity_type.HasField('/'+target_field.data):
+          print('Invalid links field target: ', target_field)
+          return False
+        # check the units are matching
+        valid_units_src = source_entity_instance.universe\
+          .GetUnitsMapByMeasurement(source_field.data)
+        valid_units_target = self .universe\
+          .GetUnitsMapByMeasurement(target_field.data)
+        # when one is None and not the other
+        if valid_units_src != valid_units_target:
+          print('Links dont have the same units: ',
+                valid_units_src, valid_units_target)
+          print('Links error for the following fields (source, target): '
+                , source_field, target_field)
+          return False
+        if valid_units_src is not None and \
+            set(valid_units_src) != set(valid_units_target):
+            print('Links dont have the same units: ',
+                  valid_units_src, valid_units_target)
+            print('Links error for the following fields (source, target): '
+                  , source_field, target_field)
+            return False
+        # check the states are matching
+        valid_states_src = source_entity_instance.universe\
+          .GetStatesByField(source_field.data)
+        valid_states_target = self.universe\
+          .GetStatesByField(target_field.data)
+        # when one is None and not the other
+        if valid_units_src != valid_units_target:
+          print('Links dont have the same states: '
+                , valid_states_src, valid_states_target)
+          print('Links error for the following fields (source, target): '
+                , source_field, target_field)
+          return False
+        if valid_states_src is not None and \
+            set(valid_states_src) != set(valid_states_target):
+          print('Links dont have the same states: ',
+                valid_states_src, valid_states_target)
+          print('Links error for the following fields (source, target): '
+                , source_field, target_field)
+          return False
+      # check if the rest of the links not included are optional
```

Comment:
this is incorrect.  It's perfectly valid for a source to not have all its required fields linked out.  

what you want to check is that the TARGET (which, conveniently, is where we define links) has all its required fields linked.